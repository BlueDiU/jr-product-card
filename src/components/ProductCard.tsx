import { useProduct } from '../hooks/useProduct';

import styles from '../styles/styles.module.css';

import React, { createContext, CSSProperties } from 'react';
import {
  InitialValues,
  IOnchangeArgs,
  IProduct,
  IProductCardHandlers,
  IProductContextProps,
} from '../interfaces/interfaces';

import { ProductTitle } from './ProductTitle';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';

export const ProductContext = createContext({} as IProductContextProps);
const { Provider } = ProductContext;

export interface IProps {
  product: IProduct;
  // children?: ReactElement | ReactElement[];
  children: (args: IProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: IOnchangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export function ProductCard({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: IProps) {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider value={{ counter, increaseBy, product, maxCount }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          maxCount,
          product,
          isMaxCountReached,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
}

ProductCard.Title = ProductTitle;
ProductCard.Image = ProductImage;
ProductCard.Buttons = ProductButtons;
