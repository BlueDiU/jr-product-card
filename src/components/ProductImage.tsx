import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';

import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';

export interface IProps {
  className?: string;
  img?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ img, className }: IProps) => {
  const { product } = useContext(ProductContext);

  const ImageToShow = img ? img : product.img ?? noImage;

  return (
    <img
      src={ImageToShow}
      alt="Product"
      className={`${styles.productImg} ${className}`}
      style={styles}
    />
  );
};
