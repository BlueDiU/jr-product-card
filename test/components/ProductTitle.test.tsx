// import * as React from 'react';
import React from 'react';
import renderer from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../data/products';

describe('ProductTitle', () => {
  test('must show the component correctly with the custom title', () => {
    const wrapper = renderer.create(<ProductTitle title="custom-product" />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('must show the component with the product name', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>{() => <ProductTitle />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
