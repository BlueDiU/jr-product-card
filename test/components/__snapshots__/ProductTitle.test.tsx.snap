// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle must show the component correctly with the custom title 1`] = `
<span
  className="productDescription undefined"
>
  custom-product
</span>
`;

exports[`ProductTitle must show the component with the product name 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug Meme - no image
  </span>
</div>
`;
